# MC316 Final Exam – Part B: PYTHON

* You will have 1 hour and 40 minutes to finish the test.
* This part includes 10 questions. Each question is worth 15 points.
* You get extra 30 points for full answers.
* Please read the instructions carefully. 

## I. SET UP


### Import pandas library for use in scripting. 

In [9]:
import pandas as pd

### Import the datasets and name them.
For this test, you will analyze revenue and expenditure data released by Kansas Department of Administration for the fiscal year 2018. While the original data can be directly downloaded [here](http://kanview.ks.gov/DataDownload.aspx), for your convenience, the instructor had them uploaded to a server at the following urls. 

#### For Kansas State University:
1. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv
2. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv

#### For University of Kansas:
3. Expenditure data: https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv
4. Revenue data: https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv

You will need to import all 4 datasets and name them in an easy way to call out later.

*Please refer to the [Glossary](http://kanview.ks.gov/Glossary.aspx#f) for data description.

In [11]:
KSUexpen = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

In [12]:
KSUrev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_367_2018.csv")

In [13]:
KUexpen = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

In [14]:
KUrev = pd.read_csv("https://datajournalism.tech/wp-content/uploads/2019/05/RevData_682_2018.csv")

## II. REQUIRED QUESTIONS
* Make sure you type the answers in the cells right below the codes.
* If you need to add extra cells, click the + button in the toolbar up the top. 

### Q1: What is the total revenue recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total revenue recorded for University of Kansas in the same year?

In [28]:
KSUrev.info("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_367_2018.csv")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1910 entries, 0 to 1909
Data columns (total 22 columns):
Fiscal_Year                        1910 non-null int64
Fiscal_Quarter                     1910 non-null int64
Business_Unit                      1910 non-null int64
Business_Unit_Descr                1910 non-null object
GAAP_Fund_Type                     1910 non-null int64
GAAP_Fund_Type_Descr               1910 non-null object
Fund_Code                          1910 non-null int64
Fund_Code_Descr                    1910 non-null object
Budget_Ref                         1910 non-null int64
Budget_Ref_Descr                   1910 non-null object
Program_Code                       1910 non-null int64
Program_Code_Descr                 1910 non-null object
State_Function                     1910 non-null int64
State_Function_Descr               1910 non-null object
Primary_Account_Code               1910 non-null int64
Primary_Account_Code_Descr         1910 non-null object
Interm

In [29]:
KUrev.info("https://datajournalism.tech/wp-content/uploads/2019/05/ExpData_682_2018.csv")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1077 entries, 0 to 1076
Data columns (total 22 columns):
Fiscal_Year                        1077 non-null int64
Fiscal_Quarter                     1077 non-null int64
Business_Unit                      1077 non-null int64
Business_Unit_Descr                1077 non-null object
GAAP_Fund_Type                     1077 non-null int64
GAAP_Fund_Type_Descr               1077 non-null object
Fund_Code                          1077 non-null int64
Fund_Code_Descr                    1077 non-null object
Budget_Ref                         1077 non-null int64
Budget_Ref_Descr                   1077 non-null object
Program_Code                       1077 non-null int64
Program_Code_Descr                 1077 non-null object
State_Function                     1077 non-null int64
State_Function_Descr               1077 non-null object
Primary_Account_Code               1077 non-null int64
Primary_Account_Code_Descr         1077 non-null object
Interm

In [61]:
KSUrev.Total.max()

91090183.66

In [62]:
KUrev.Total.max()

124870966.43

<b>Answer:</b> After adding up the total revenue for Kansas State University it made 91090183 in revenue compared to The University of Kansas that made 124870966 meaning KU made more in revenue in the fiscal year of 2018.

### Q2: What is the total expense recorded for Kansas State University in the fiscal year of 2018? How does it compare to the total expense recorded for University of Kansas in the same year?

In [59]:
KSUexpen.Total.sum()

1018584373.02

In [60]:
KUexpen.Total.sum()

990238875.6099999

<b>Answer:</b> The total expense recorded for KSU in the Fiscal year of 2018 was 1018584373 compared to KU's which was 990238875. KSU in this instance had more expenses than KU in the fiscal year of 2018.

### Q3: Where did the two universities' primary revenues come from? 
*Refer to the description [here](http://kanview.ks.gov/Revenues/AcctCodeRevenues.aspx) to write your answer.

In [76]:
KSUrev.Primary_Account_Code_Descr.value_counts().reset_index()

index  Primary_Account_Code_Descr
0  AGENCY EARNINGS                                        637
1  REV FROM USE OF MONEY & PROP                           354
2  REVENUE TRANSFERS                                      341
3  NON REVENUE RECEIPTS                                   259
4  OTHER REVENUE                                          160
5  GIFT DONATION FEDERAL GRANTS                           152
6  TAXES                                                    7

In [77]:
KUrev.Primary_Account_Code_Descr.value_counts().reset_index()

index  Primary_Account_Code_Descr
0  AGENCY EARNINGS                                        376
1  REV FROM USE OF MONEY & PROP                           251
2  NON REVENUE RECEIPTS                                   157
3  REVENUE TRANSFERS                                      153
4  OTHER REVENUE                                           57
5  TAXES                                                   45
6  GIFT DONATION FEDERAL GRANTS                            38

<b>Answer:</b> KSU had Agency Earnings as their primary revenue source at 637 as well did KU as their primary revenue source at 376

### Q4: What was the biggest source of revenues for Kansas State University? Compare to the one of University of Kansas.

Remove hashtag and run the following code to suppress scientific notation:

In [71]:
pd.set_option('display.float_format', lambda x: '%.3f' % x) 

In [93]:
KSUrev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                 72982844.780
GIFT DONATION FEDERAL GRANTS    11511087.710
NON REVENUE RECEIPTS            38810748.620
OTHER REVENUE                   21647825.100
REV FROM USE OF MONEY & PROP    88209447.500
REVENUE TRANSFERS              237422209.080
TAXES                           85937809.510
Name: Total, dtype: float64

In [95]:
KUrev.groupby("Primary_Account_Code_Descr").Total.sum()

Primary_Account_Code_Descr
AGENCY EARNINGS                  408409337.320
GIFT DONATION FEDERAL GRANTS     193836668.320
NON REVENUE RECEIPTS              69753247.670
OTHER REVENUE                     68241930.500
REV FROM USE OF MONEY & PROP      40950779.540
REVENUE TRANSFERS                 26013855.400
TAXES                              1562083.800
Name: Total, dtype: float64

<b>Answer:</b> The biggest source of revenues for KSU was Revenue Transfers compared to KU's which was 1,077 making a 833 difference between the two biggest source of revenues for the Universities.

### Q5: What accounts contributed to Kansas State University's biggest source of revenues?
If you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [2]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

After cleaning the spaces, you can start filtering your datasets to find the answer for Q5.

In [92]:
KSUrev["Primary_Account_Code_Descr"] = KUrev["Primary_Account_Code_Descr"].str.strip()

<b>Answer:</b> 

### Q6: Where did the money go? (Answer for both universities)
*Refer to the description [here] (http://kanview.ks.gov/Expenditures/AcctCodeExpenditures.aspx) to write your answer.

In [86]:
KSUexpen.Primary_Account_Code_Descr.value_counts().reset_index()

index  Primary_Account_Code_Descr
0  CONTRACTUAL SERVICES                                 10561
1  SALARIES AND WAGES                                    7780
2  COMMODITIES                                           3740
3  CAPITAL OUTLAY                                        1173
4  GRANTS CLAIMS SHARED REVENUE                           534
5  NONEXPENSE ITEMS                                       129
6  EXPENSE TRANSFERS                                      106
7  DEBT SERVICE                                            71

In [87]:
KUexpen.Primary_Account_Code_Descr.value_counts().reset_index()

index  Primary_Account_Code_Descr
0  CONTRACTUAL SERVICES                                  4322
1  SALARIES AND WAGES                                    3623
2  COMMODITIES                                           1605
3  CAPITAL OUTLAY                                         737
4  NONEXPENSE ITEMS                                        81
5  GRANTS CLAIMS SHARED REVENUE                            75
6  EXPENSE TRANSFERS                                       66
7  DEBT SERVICE                                            44

<b>Answer:</b> Where the money for both Universities went was Contractual Services and KSU had 10,561 for contractual services & KU had 4,322 for contractual services.

### Q7: What costs two universities the most? 

In [102]:
KSUexpen.Primary_Account_Code_Descr.value_counts().reset_index()

index  Primary_Account_Code_Descr
0  CONTRACTUAL SERVICES                                 10561
1  SALARIES AND WAGES                                    7780
2  COMMODITIES                                           3740
3  CAPITAL OUTLAY                                        1173
4  GRANTS CLAIMS SHARED REVENUE                           534
5  NONEXPENSE ITEMS                                       129
6  EXPENSE TRANSFERS                                      106
7  DEBT SERVICE                                            71

<b>Answer:</b> The two universities were cost the most by Agency Earnings

### Q8: Calculate the percentage of this cost over each university's total expense.

<b>Answer:</b> 

### Q9: Breakdown the cost you found in Q7 for both universities.
Again, if you run into a situation where a filter function returns 0 results, you may want to check if there are leading and trailing spaces in the column. To clean the leading and trailing spaces, you will need to remove the hashtag, revise the names, and run the following code:

In [3]:
#yourdataframe["yourcolumnname"] = yourdataframe["yourcolumnname"].str.strip()

<b>Answer:</b>

### Q10: Compare the breakdown costs from Q9 side by side.
When you merge the two subsets, make sure to add the attribute <b>how ="outer"</b> so that you can keep every item from both sets.
*For example: merge = pd.merge(df1, df2, how="outer", on="columnname")

<b>Answer:</b>

## III. BONUS QUESTIONS
* Each question's worth 15 points.


### B1: Which university made a profit in the fiscal year of 2018?

<b>Answer:</b> Kansas State University

### B2: Which software program do you like the most in this course? Why?

<b>Answer:</b> My most favorite software program was R Studio in this course. This was my favorite even though we didn't get tested over it because of the time constraint but I learned most because of the process and way of functions that we had to insert and were able to recieve a clear outcome with what we were either doing wrong or doing correct!

### END OF THE TEST